envoy/server.go

```diff
@@ -0,0 +1,437 @@
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+	var listeners []cache.Resource
+
+	s.state.RLock()
+	defer s.state.RUnlock()
+
+	svcs := s.state.ByService()
+	// Loop over all the services by service name
+	for _, endpoints := range svcs {
+		if len(endpoints) < 1 {
+			continue
+		}
+
+		var svc *service.Service
+		// Find the first alive service and use that as the definition.
+		// If none are alive, we won't open the port.
+		for _, endpoint := range endpoints {
+			if endpoint.IsAlive() {
+				svc = endpoint
+				break
+			}
+		}
+
+		if svc == nil {
+			continue
+		}
+
+		// Loop over the ports and generate a named listener for
+		// each port.
+		for _, port := range svc.Ports {
+			// Only listen on ServicePorts
+			if port.ServicePort < 1 {
+				continue
+			}
+
+			listener, err := s.MakeListenerFromService(svc, uint32(port.ServicePort))
+			if err != nil {
+				return nil, fmt.Errorf("failed to create listener from service: %s", err)
+			}
+			listeners = append(listeners, listener)
+		}
+	}
+
+	log.Debugf("Created %d Envoy listeners", len(listeners))
+
+	return listeners, nil
+}
+
+// Format an Envoy service name from our service name and port
+func SvcName(name string, port uint32) string {
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+	// TODO: Move this code in a common module and de-duplicate in sidecarhttp/envoy_api.go
+	// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+	return fmt.Sprintf("%s%s%d", name, ServiceNameSeparator, port)
+}
+
+// MakeListenerFromService creates an Envoy listener from a service instance
+func (s *Server) MakeListenerFromService(svc *service.Service, port uint32) (cache.Resource, error) {
+	apiName := SvcName(svc.Name, port)
+
+	var connectionManagerName string
+	var connectionManager proto.Message
+	switch svc.ProxyMode {
+	case "http":
+		connectionManagerName = wellknown.HTTPConnectionManager
+
+		connectionManager = &hcm.HttpConnectionManager{
+			StatPrefix: "ingress_http",
+			HttpFilters: []*hcm.HttpFilter{{
+				Name: wellknown.Router,
+			}},
+			RouteSpecifier: &hcm.HttpConnectionManager_RouteConfig{
+				RouteConfig: &api.RouteConfiguration{
+					VirtualHosts: []*route.VirtualHost{{
+						Name:    apiName,
+						Domains: []string{"*"},
+						Routes: []*route.Route{{
+							Match: &route.RouteMatch{
+								PathSpecifier: &route.RouteMatch_Prefix{
+									Prefix: "/",
+								},
+							},
+							Action: &route.Route_Route{
+								Route: &route.RouteAction{
+									ClusterSpecifier: &route.RouteAction_Cluster{
+										Cluster: apiName,
+									},
+									// TODO: No timeout? This is what we did for the REST API, but not
+									// sure if it's the right thing to do.
+									Timeout: &duration.Duration{},
+									// TODO: Do we need to set this? We were setting it in the REST API,
```

Comment:
It was definitely needed at one point. I don't remember the specifics. If it works without it on V2, then great.